File: web/src/lib/api/storage.ts
import { del, get, patch, post, put } from '@/lib/http'
import type {
  CreateStorageConfigRequest,
  DeleteStorageConfigResponse,
  DriverSchemaRequest,
  DriverSchemaResponse,
  PageData,
  StorageConfig,
  StorageDriver,
  TestStorageConfigResponse,
  UpdateStorageConfigRequest,
  UpdateStorageSecretsRequest,
  UpdateStorageSecretsResponse,
} from '@/types/api'

/**
 * 存储驱动（API.md §3）。**全部端点需要 admin**。
 *
 * 两条边界（DESIGN.md §5.3）：
 *  - 同一驱动类型可有多条实例（D64）：列表展示 `Name`，内部一律用 `UID`
 *  - **响应中绝不含密钥**（D78）：编辑时密钥字段留空 = 不修改
 */

export interface StorageConfigListParams {
  Page?: number
  PageSize?: number
  Enabled?: boolean
  Type?: string
  Keyword?: string
}

export const storageApi = {
  /** 可用驱动类型 + **服务端求值后的** 配置字段 schema（来自 agent 探测）。 */
  drivers(): Promise<{ Drivers: StorageDriver[] }> {
    return get<{ Drivers: StorageDriver[] }>('/storage/drivers')
  },

  /**
   * 按当前表单值**重新求值** schema（处理 `DependsOn` 联动）。
   *
   * ⚠️ 联动必须回到服务端：`choices` / `default` 可能是插件函数，
   * 前端**永不执行插件代码**（DESIGN.md §7.3）。
   */
  driverSchema(payload: DriverSchemaRequest): Promise<DriverSchemaResponse> {
    return post<DriverSchemaResponse>('/storage/drivers/schema', payload)
  },

  list(params: StorageConfigListParams = {}): Promise<PageData<StorageConfig>> {
    return get<PageData<StorageConfig>>('/storage/configs', { params })
  },

  get(uid: string): Promise<StorageConfig> {
    return get<StorageConfig>(`/storage/configs/${encodeURIComponent(uid)}`)
  },

  create(payload: CreateStorageConfigRequest): Promise<StorageConfig> {
    return post<StorageConfig>('/storage/configs', payload)
  },

  /** 更新元数据与模板（**不含密钥**；`PicgoConfigName` 创建后只读）。 */
  update(uid: string, payload: UpdateStorageConfigRequest): Promise<StorageConfig> {
    return patch<StorageConfig>(`/storage/configs/${encodeURIComponent(uid)}`, payload)
  },
  /** 单独更新凭据（与元数据分离，D78）。**只提交需变更的字段**。 */
  updateSecrets(
    uid: string,
    payload: UpdateStorageSecretsRequest,
  ): Promise<UpdateStorageSecretsResponse> {
    return put<UpdateStorageSecretsResponse>(
      `/storage/configs/${encodeURIComponent(uid)}/secrets`,
      payload,
    )
  },

  /** 删除。`force` 为真时忽略「有图片引用」的冲突。 */
  remove(uid: string, force = false): Promise<DeleteStorageConfigResponse> {
    return del<DeleteStorageConfigResponse>(`/storage/configs/${encodeURIComponent(uid)}`, {
      params: { Force: force },
    })
  },

  /** 置为全局默认（同事务把其他行置 false，并让 agent 切换当前上传器）。 */
  activate(uid: string): Promise<StorageConfig> {
    return post<StorageConfig>(`/storage/configs/${encodeURIComponent(uid)}/activate`)
  },

  /** 连通性测试（**HTTP 仍为 200**，看 `Ok` 字段判断结果）。 */
  test(uid: string): Promise<TestStorageConfigResponse> {
    return post<TestStorageConfigResponse>(`/storage/configs/${encodeURIComponent(uid)}/test`)
  },
}

/** 供上传页等处复用的「拿默认存储」便捷函数。 */
export async function fetchDefaultStorageConfig(): Promise<StorageConfig | undefined> {
  const page = await storageApi.list({ Page: 1, PageSize: 100, Enabled: true })
  return page.Items.find((item) => item.IsDefault) ?? page.Items[0]
}

File: server/internal/handler/storage.go
package handler

import (
	"log/slog"
	"strings"

	"github.com/gin-gonic/gin"

	"github.com/YeqingKy/PicGo-Web/server/internal/middleware"
	"github.com/YeqingKy/PicGo-Web/server/internal/response"
	"github.com/YeqingKy/PicGo-Web/server/internal/service"
)

// StorageHandler 处理存储驱动配置端点（/storage/**）。
//
// 权限：**全部端点需要 admin**（D4：存储驱动只有管理员能配置）。
// 由 RegisterBusiness 统一挂 `RequireAuth + RequirePasswordChanged + RequireAdmin`。
type StorageHandler struct {
	storage *service.StorageService
	log     *slog.Logger
}

// NewStorageHandler 构造。
func NewStorageHandler(storage *service.StorageService, log *slog.Logger) *StorageHandler {
	return &StorageHandler{storage: storage, log: log}
}

// Register 注册路由。
func (h *StorageHandler) Register(g *gin.RouterGroup) {
	g.GET("/drivers", h.ListDrivers)
	g.POST("/drivers/schema", h.DriverSchema)

	g.GET("/configs", h.List)
	g.POST("/configs", h.Create)
	g.GET("/configs/:UID", h.Get)
	g.PATCH("/configs/:UID", h.Update)
	g.PUT("/configs/:UID/secrets", h.UpdateSecrets)
	g.DELETE("/configs/:UID", h.Delete)
	g.POST("/configs/:UID/activate", h.Activate)
	g.POST("/configs/:UID/test", h.Test)
}

// ListDrivers 处理 GET /storage/drivers。
func (h *StorageHandler) ListDrivers(c *gin.Context) {
	drivers, err := h.storage.ListDrivers(c.Request.Context())
	if err != nil {
		respondServiceError(c, h.log, err)
		return
	}
	response.OK(c, gin.H{"Drivers": drivers})
}

// DriverSchema 处理 POST /storage/drivers/schema（DependsOn 联动重求值）。
func (h *StorageHandler) DriverSchema(c *gin.Context) {
	var req struct {
		Type    string         `json:"Type"`
		Answers map[string]any `json:"Answers"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.InvalidParam(c, "请求体格式不正确")
		return
	}
	if strings.TrimSpace(req.Type) == "" {
		response.InvalidParam(c, "缺少 Type")
		return
	}

	driver, err := h.storage.DriverSchema(c.Request.Context(), req.Type, req.Answers)
	if err != nil {
		respondServiceError(c, h.log, err)
		return
	}
	response.OK(c, driver)
}

// List 处理 GET /storage/configs。
func (h *StorageHandler) List(c *gin.Context) {
	var enabled *bool
	if raw := strings.TrimSpace(c.Query("Enabled")); raw != "" {
		v := raw == "1" || strings.EqualFold(raw, "true")
		enabled = &v
	}

	items, total, err := h.storage.List(c.Request.Context(), service.StorageListInput{
		Keyword:  strings.TrimSpace(c.Query("Keyword")),
		Type:     strings.TrimSpace(c.Query("Type")),
		Enabled:  enabled,
		Page:     queryInt(c, "Page", service.DefaultPage),
		PageSize: queryInt(c, "PageSize", service.DefaultPageSize),
	})
	if err != nil {
		respondServiceError(c, h.log, err)
		return
	}
	response.Page(c, items, total,
		clampPage(queryInt(c, "Page", service.DefaultPage)),
		clampPageSize(queryInt(c, "PageSize", service.DefaultPageSize)))
}

// Get 处理 GET /storage/configs/:UID。
func (h *StorageHandler) Get(c *gin.Context) {
	view, err := h.storage.Get(c.Request.Context(), c.Param("UID"))
	if err != nil {
		respondServiceError(c, h.log, err)
		return
	}
	response.OK(c, view)
}

// Create 处理 POST /storage/configs。
func (h *StorageHandler) Create(c *gin.Context) {
	var req struct {
		Name            string         `json:"Name"`
		Type            string         `json:"Type"`
		PicgoConfigName string         `json:"PicgoConfigName"`
		Enabled         *bool          `json:"Enabled"`
		IsDefault       bool           `json:"IsDefault"`
		PathTemplate    string         `json:"PathTemplate"`
		FileTemplate    string         `json:"FileTemplate"`
		Config          map[string]any `json:"Config"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.InvalidParam(c, "请求体格式不正确")
		return
	}

	view, err := h.storage.Create(c.Request.Context(), service.CreateStorageInput{
		Name:            req.Name,
		Type:            req.Type,
		PicgoConfigName: req.PicgoConfigName,
		Enabled:         req.Enabled,
		IsDefault:       req.IsDefault,
		PathTemplate:    req.PathTemplate,
		FileTemplate:    req.FileTemplate,
		Config:          req.Config,
	}, middleware.CurrentUserUID(c), middleware.ClientIP(c), c.Request.UserAgent())
	if err != nil {
		respondServiceError(c, h.log, err)
		return
	}
	response.OK(c, view)
}

// Update 处理 PATCH /storage/configs/:UID。
//
// ⚠️ **不接受 `PicgoConfigName`**（创建后只读，D64 推论）：
// picgo 的 `createOrUpdate` 是「未命中即新建」语义，允许改名会把同一条配置裂成两条。
// 展示名要改就改 `Name`。
func (h *StorageHandler) Update(c *gin.Context) {
	var req struct {
		Name         *string `json:"Name"`
		Enabled      *bool   `json:"Enabled"`
		IsDefault    *bool   `json:"IsDefault"`
		PathTemplate *string `json:"PathTemplate"`
		FileTemplate *string `json:"FileTemplate"`

		// 显式声明以便**明确忽略**并给出提示（而不是让调用方以为改成功了）
		PicgoConfigName *string `json:"PicgoConfigName"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.InvalidParam(c, "请求体格式不正确")
		return
	}
	if req.PicgoConfigName != nil {
		response.InvalidParam(c, "PicgoConfigName 创建后不可修改（如需改名请改 Name）")
		return
	}

	view, err := h.storage.Update(c.Request.Context(), c.Param("UID"), service.UpdateStorageInput{
		Name:         req.Name,
		Enabled:      req.Enabled,
		IsDefault:    req.IsDefault,
		PathTemplate: req.PathTemplate,
		FileTemplate: req.FileTemplate,
	}, middleware.CurrentUserUID(c), middleware.ClientIP(c), c.Request.UserAgent())
	if err != nil {
		respondServiceError(c, h.log, err)
		return
	}
	response.OK(c, view)
}

// UpdateSecrets 处理 PUT /storage/configs/:UID/secrets（单独更新凭据）。
func (h *StorageHandler) UpdateSecrets(c *gin.Context) {
	var req struct {
		Config map[string]any `json:"Config"`
	}
	if err := c.ShouldBindJSON(&req); err != nil {
		response.InvalidParam(c, "请求体格式不正确")
		return
	}

	view, err := h.storage.UpdateSecrets(c.Request.Context(), c.Param("UID"), req.Config,
		middleware.CurrentUserUID(c), middleware.ClientIP(c), c.Request.UserAgent())
	if err != nil {
		respondServiceError(c, h.log, err)
		return
	}
	response.OK(c, gin.H{
		"UID":          view.UID,
		"HasSecrets":   view.HasSecrets,
		"SecretFields": view.SecretFields,
	})
}

// Delete 处理 DELETE /storage/configs/:UID。
func (h *StorageHandler) Delete(c *gin.Context) {
	force := strings.EqualFold(strings.TrimSpace(c.Query("Force")), "true") ||
		strings.TrimSpace(c.Query("Force")) == "1"

	result, err := h.storage.Delete(c.Request.Context(), c.Param("UID"), force,
		middleware.CurrentUserUID(c), middleware.ClientIP(c), c.Request.UserAgent())
	if err != nil {
		respondServiceError(c, h.log, err)
		return
	}
	response.OK(c, result)
}

// Activate 处理 POST /storage/configs/:UID/activate。
func (h *StorageHandler) Activate(c *gin.Context) {
	view, err := h.storage.Activate(c.Request.Context(), c.Param("UID"),
		middleware.CurrentUserUID(c), middleware.ClientIP(c), c.Request.UserAgent())
	if err != nil {
		respondServiceError(c, h.log, err)
		return
	}
	response.OK(c, view)
}

// Test 处理 POST /storage/configs/:UID/test。
//
// ⚠️ 测试失败时 HTTP 仍为 200（「测试本身执行成功」），
// 结果用 `Ok` / `Message` 表达 —— 见 docs/API.md §3.2。
func (h *StorageHandler) Test(c *gin.Context) {
	result, err := h.storage.Test(c.Request.Context(), c.Param("UID"),
		middleware.CurrentUserUID(c), middleware.ClientIP(c), c.Request.UserAgent())
	if err != nil {
		respondServiceError(c, h.log, err)
		return
	}
	response.OK(c, result)
}
